Core/include/Acts/SpacePointFormation/detail/SpacePointBuilder.ipp

```diff
@@ -0,0 +1,673 @@
+        measurementPairs) const {
+  // Return if no Measurements are given in a vector
+  if (measurementsFront.empty() || measurementsBack.empty()) {
+    return;
+  }
+  // Declare helper variables
+  double currentDiff;
+  double diffMin;
+  unsigned int measurementMinDist;
+
+  // Walk through all Measurements on both surfaces
+  for (unsigned int iMeasurementsFront = 0;
+       iMeasurementsFront < measurementsFront.size(); iMeasurementsFront++) {
+    // Set the closest distance to the maximum of double
+    diffMin = std::numeric_limits<double>::max();
+    // Set the corresponding index to an element not in the list of Measurements
+    measurementMinDist = measurementsBack.size();
+    for (unsigned int iMeasurementsBack = 0;
+         iMeasurementsBack < measurementsBack.size(); iMeasurementsBack++) {
+      auto [gposFront, gcovFront] =
+          globalCoords(gctx, *(measurementsFront[iMeasurementsFront]));
+      auto [gposBack, gcovBack] =
+          globalCoords(gctx, *(measurementsBack[iMeasurementsBack]));
+
+      currentDiff = detail::differenceOfMeasurementsChecked(
+          gposFront, gposBack, m_config.vertex, m_config.diffDist,
+          m_config.diffPhi2, m_config.diffTheta2);
+
+      // Store the closest Measurements (distance and index) calculated so far
+      if (currentDiff < diffMin && currentDiff >= 0.) {
+        diffMin = currentDiff;
+        measurementMinDist = iMeasurementsBack;
+      }
+    }
+
+    // Store the best (=closest) result
+    if (measurementMinDist < measurementsBack.size()) {
+      std::pair<const Measurement*, const Measurement*> measurementPair;
+      measurementPair = std::make_pair(measurementsFront[iMeasurementsFront],
+                                       measurementsBack[measurementMinDist]);
+      measurementPairs.push_back(measurementPair);
+    }
+  }
+}
+
+template <typename spacepoint_t>
+template <template <typename...> typename container_t>
+void SpacePointBuilder<spacepoint_t>::calculateDoubleHitSpacePoints(
+    const GeometryContext& gctx,
+    const std::vector<std::pair<const Measurement*, const Measurement*>>&
+        measurementPairs,
+    std::back_insert_iterator<container_t<spacepoint_t>> spacePointIt) const {
+  // Source of algorithm: Athena, SiSpacePointMakerTool::makeSCT_SpacePoint()
+
+  detail::SpacePointParameters spaPoPa;
+
+  // Walk over every found candidate pair
+  for (const auto& mp : measurementPairs) {
+    // Calculate the ends of the SDEs
+    const auto& ends1 = endsOfStrip(gctx, *(mp.first));
+    const auto& ends2 = endsOfStrip(gctx, *(mp.second));
+
+    spaPoPa.q = ends1.first - ends1.second;
+    spaPoPa.r = ends2.first - ends2.second;
+
+    const auto& slink_front = getSourceLink(*(mp.first));
+    const auto& slink_back = getSourceLink(*(mp.second));
+
+    boost::container::static_vector<const SourceLink*, 2> slinks;
+    slinks.emplace_back(slink_front);
+    slinks.emplace_back(slink_back);
+
+    double resultPerpProj;
```

Comment:
this is only used inside `if (m_config.usePerpProj)`. I'd move that there as 
```
double resultPerpProj = detail::calcPerpendicularProjection(
          ends1.first, ends2.first, spaPoPa.q, spaPoPa.r);
```

Reply:
fixed